src/psyclone/psyir/nodes/reference.py

```diff
@@ -94,6 +95,17 @@ def is_array(self):
         '''
         return self.symbol.is_array
 
+    @property
+    def is_character(self):
+        '''
+        :returns: True if this reference is a character, otherwise False.
+        :rtype: bool
+
+        '''
+        return (
+            self.symbol.datatype._intrinsic == ScalarType.Intrinsic.CHARACTER
```

Comment:
Note that you would also need to use: `self.datatype` for References anyway because not all symbols have a datatype attribute (alternatively you would need the proper checks here).

Similarly, not all datatypes have "intrinsic" (you can check it with `hasattr(self.datatype, "intrinsic")`).

Which raises the question about what should it return if "we don't know if its character or not". For example for:
```
subroutine test
    use other_module, only: a, b, c
    a = b + c
end subroutine
```

- One solution would be to return True or False and rename the method as could_be_character / guaranteed_to_be_character . This would be fine but it is unclear to me what is the best as I can see scenarios for both.

- Another solution would be to return something different or raise an Error. But this leads to code that always needs the "checks" or "try excepts" around it, which is easy to forget.

- A third solution is to add an argument to chose what to return. e.g. `reference.is_character(unknowns_as=True)`. The drawback is that it is not a "property" anymore, but the argument can be omitted if we agree in a good default.

What do you think?

Reply:
Thinking a bit more about this I don't like the "return something different" because this "something" needs to be imported or be None. While none/null would be the prefered solution in other languages, in python None evaluates to False in conditions, so `if reference.is_character:` will behave as returning False for Unknowns which could be surprising if not familiar with the API. So I am leaning towards the third option because `if reference.is_character():` will work as expected for users until it finds a complex case, in that case it would raise and error that can inform about the unkonws_as optional argument to the user, and there is nothing unexpected in its behaviour.

This is consistent with the "otherwise" parameter that we introduced in the symbol table and maybe could be a rule/style for similar methods. What do @arporter @hiker think?

Reply:
I like the third option too. In #2661 I have also hit the problem of wanting to query something that can be True, False or Unknown. I've temporarily introduced a `Fuzzy` enum that contains values for TRUE, FALSE and MAYBE but I'm unsure yet whether this is a good solution.

Reply:
Thanks. I implemented what I think you mean in 772d3ff1726e96db078b7d9fbc4e7c0a98e2f3a8, where `unknown_as` is a `bool` that gives the return value if it can't be determined whether the `DataNode` is a character. However, you mention "error by default" so should the flag instead be about whether to raise an error if the value is unknown?

Reply:
What I meant is for the `unknown_as` to be of type `Optional[bool]` and default to `=None`

In the implementation:

- if it is none and an unresolvable case is found, raise and error saying something like "is_character could not resolve whether the expression {self.debug_string()} operates on characters. Note that 'is_character' has the optional argument 'unknown_as' to provide a fallback value to return in this case."

- If it is not none, just return "unknown_as" as you do now.

Reply:
Ah I see, thanks. Addressed in ae82fd765c5c5a3b1bde00c660b617ef701a83ec.